src/Grpc.AspNetCore.Server/Internal/HttpContextSerializationContext.cs

```diff
@@ -0,0 +1,232 @@
+
+            var headerData = pipeWriter.GetSpan(HeaderSize);
+
+            // Compression flag
+            headerData[0] = compress ? (byte)1 : (byte)0;
+
+            // Message length
+            BinaryPrimitives.WriteUInt32BigEndian(headerData.Slice(1), (uint)length);
+
+            pipeWriter.Advance(HeaderSize);
+        }
+
+        public override IBufferWriter<byte> GetBufferWriter()
+        {
+            switch (_state)
+            {
+                case InternalState.Initialized:
+                    // When writing directly to the buffer the header with message size needs to be written first
+                    if (DirectSerializationSupported)
+                    {
+                        Debug.Assert(_payloadLength != null, "A payload length is required for direct serialization.");
+
+                        EnsureMessageSizeAllowed(_payloadLength.Value);
+
+                        WriteHeader(ResponseBufferWriter, _payloadLength.Value, compress: false);
+                    }
+
+                    _state = InternalState.IncompleteBufferWriter;
+                    goto case InternalState.IncompleteBufferWriter;
+                case InternalState.IncompleteBufferWriter:
+                    return DirectSerializationSupported
+                        ? (IBufferWriter<byte>)ResponseBufferWriter
+                        : _bufferWriter ??= new ArrayBufferWriter<byte>();
+                default:
+                    ThrowInvalidState(_state);
+                    return default!;
+            }
+        }
+
+        private void EnsureMessageSizeAllowed(int payloadLength)
+        {
+            if (payloadLength > _serverCallContext.MethodContext.MaxSendMessageSize)
+            {
+                throw new RpcException(SendingMessageExceedsLimitStatus);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void ThrowInvalidState(InternalState state)
+            => throw new InvalidOperationException(state.ToString());
+
+        public override void Complete()
+        {
+            switch (_state)
+            {
+                case InternalState.IncompleteBufferWriter:
+                    _state = InternalState.CompleteBufferWriter;
+
+                    if (!DirectSerializationSupported)
+                    {
+                        Debug.Assert(_bufferWriter != null, "Buffer writer has been set to get to this state.");
+
+                        var data = _bufferWriter.WrittenSpan;
+
+                        GrpcServerLog.SerializedMessage(_serverCallContext.Logger, typeof(object), data.Length);
+                        WriteMessage(data);
+                    }
+                    else
+                    {
+                        GrpcServerLog.SerializedMessage(_serverCallContext.Logger, _serverCallContext.MethodContext.ResponseType, _payloadLength.GetValueOrDefault());
```

Comment:
I chose not to because this code will only be overridden by frameworks. They should test they are doing the right thing themselves using unit/functional tests.

Checking every serialization feels wasteful.

Reply:
I'm going to commit this as it is. If you think this is important, or there are bug reports around this issue after release, then it can be added later by wrapping the buffer writer returned.